pkg/virt-controller/watch/snapshot/snapshot.go

```diff
@@ -990,6 +1013,42 @@ func (ctrl *VMSnapshotController) checkOnlineSnapshotting(vmSnapshot *snapshotv1
 	return (vmRunning || exists), nil
 }
 
+func (ctrl *VMSnapshotController) handleVMSnapshotDeadlineExceeded(vmSnapshot *snapshotv1.VirtualMachineSnapshot, source snapshotSource) error {
+	if vmSnapshotFailed(vmSnapshot) && vmSnapshot.DeletionTimestamp == nil {
+		return nil
+	}
+
+	if source != nil {
+		if _, err := source.Unlock(); err != nil {
+			return err
+		}
+	}
+
+	vmSnapshotCpy := vmSnapshot.DeepCopy()
+	err := ctrl.unfreezeGuestFSIfNeeded(vmSnapshotCpy)
+	if err != nil {
+		reason := "Failed unfreezing guest FS"
+		ctrl.updateVMSnapshotError(vmSnapshotCpy, reason)
+		return err
+	}
+
+	if vmSnapshot.DeletionTimestamp != nil {
+		return ctrl.cleanupVMSnapshot(vmSnapshotCpy)
+	}
+
+	log.Log.Errorf("vm Snapshot %s deadline exceeded, marking as failed", vmSnapshot.Name)
+	vmSnapshotCpy.Status.Phase = snapshotv1.Failed
+	updateSnapshotCondition(vmSnapshotCpy, newProgressingCondition(corev1.ConditionFalse, vmSnapshotDeadlineExceededError))
+	updateSnapshotCondition(vmSnapshotCpy, newFailureCondition(corev1.ConditionTrue, vmSnapshotDeadlineExceededError))
+	if !reflect.DeepEqual(vmSnapshot, vmSnapshotCpy) {
+		if _, err := ctrl.Client.VirtualMachineSnapshot(vmSnapshotCpy.Namespace).Update(context.Background(), vmSnapshotCpy, metav1.UpdateOptions{}); err != nil {
+			return err
+		}
+	}
+
```

Comment:
I think updateSnapshotStatus should be the place that handles the status changes.

This `handleVMSnapshotDeadlineExceeded` function should understand what actions need to take place as a result of a Deadline being exceeded, and the updateSnapshotStatus should be in charge of moving the snapshot to a failed state.

Reply:
I think I understand what you are saying but I'm not sure that we are too far off from what you describe.  There is now one additional place where a snapshot is updated.  There is only one other place.  I don't think the code is hard to follow, but of course I think that